whiteboard + hostrequires beaker support, with docs and examples updates

seandst opened the pull request:
This PR adds support for the beaker job whiteboard to beaker topologies, and improves support for the beaker API's `hostRequires` filtering. Docs related to the hostrequires configuration have been added, and I've also updated the beaker examples.

closes #182
closes #196

seandst commented on linchpin/defaults/schemas/schema_v4.json:
```diff
@@ -108,8 +108,35 @@
+                                             "allOf": [
+                                                 {"required": ["force", "tag"]}
+                                             ]
+                                         },
```
This schema bit is attempting to reflect the mutually exclusive relationship between the use of `force:` under `hostrequires`, and the use of `tag:`, since using `force:` means that any `tag:` filters are ignored by beaker. I have no idea if it works, and would love some pointers on simple ways to actually test this scheme validation against the topologies I've written during the development and testing of this PR.

herlo replied:
We have written a schemacheck.yml, included at the same level as site.yml, which you could use to check your schema. You use it something like this: 

```bash
ansible-playbook -vvvv linchpin/provision/schemacheck.yml -e \ 
"topology='/home/herlo/Projects/linchpin/linchpin/examples/topologies/dummy_cluster.yml'" \ 
-e 'schema=../defaults/schemas/schema_v4.json'
```

herlo replied:
@samvarankashyap any opinions on this?

seandst replied:
schemacheck looks perfect, I'll run some examples through the schema and see how I did. :)

seandst replied:
Alright, "bad" values are rejected, but the mutual exclusion stuff doesn't work. I'll fix it up.

seandst replied:
> Is there any specific reason you used allof line 115 of schemav4 instead of oneof ?

It's a under a "not", and I'm trying to express than you can have anyOf objects with a "tag" key or objects with a "force" key, and also that you can not have allOf objects with a "tag" key and objects with a "force" key. It only makes sense to use "force" once, but it's acceptable to have multiple tags.

oneOf might do what I want here, but I think I'm going to break this out into its own issue, and pull the mutual exclusion stuff out of this PR. The included docs for beaker topologies, at least, explain that any included "tag" objects will be ignored if "force" is used.

samvarankashyap commented:
@seandst @herlo : Looks good to me 
@seandst Is there any specific reason you used allof line 115 of schemav4 instead of oneof ?

seandst commented:
Alright, I removed the mutual exclusion validation piece. "bad" objects under hostrequires are indeed rejected as invalid, which I think it good enough for me. If anyone thinks there's value in adding the mutual exclusion to linchpin validation I'll open an issue to deal with it. In hindsight, I don't think it adds value, because both the beaker client and the beaker API allow it, and just ignore any other host filters if "force" is set. Since, in my mind, the validation's main goal is to ultimately ensure we're sending good data to providers, the schema I've added is good enough.
